content/blog/2021-07-13-hyperparam-tuning.md

```diff
@@ -0,0 +1,294 @@
+pictureComment: Tuning Hyperparameters with Reproducible Experiments
+author: milecia_mcgregor
+commentsUrl: https://discuss.dvc.org/t/utilizing-custom-git-references-in-dvc/727
+tags:
+  - MLOps
+  - Git
+  - Experiments
+  - Reproducibility
+---
+
+## Intro
+
+When you're starting to build a new machine learning model and you're deciding
+on the model architecture, there are a number of issues that arise. You have to
+monitor every code change you make, note any differences in your data, and keep
+up with hyperparameter value updates.
+
+This means it's important for you to be able to track all of these changes so
+you can reproduce your experiments without wondering which changes gave you the
+best model. Having reproducibility in your training gives you the ability to go
+back to any point in your process to see which changes gave you the best
+results.
+
+In this tutorial, we're going to go through an example of hyperparameter tuning
+with reproducibility using DVC. You can add this to any existing project you're
+working on or start from a fresh project.
+
+## Background on Hyperparameters
+
+Before jumping straight into training and experiments, let's briefly go over
+some background on hyperparameters. Hyperparameters are the values that define
+your model. This includes things like the number of layers in a neural network
+or the learning rate for gradient descent.
+
+These parameters are different from model parameters because we can't get them
+from training our model. They are used to _create_ the model we train with.
+Optimizing these values means running training steps for different kinds of
+models to see how accurate the results are. We can get the best model from
+iterating through different hyperparameters values and seeing how they effect
+our accuracy.
+
+That's why we do hyperparameter tuning. There are a couple common methods that
+we'll do some code examples with: grid search and random search.
+
+## Tuning with DVC
+
+Let's start by talking about DVC a bit because we'll be using it to add
+reproducibility to our tuning process. This is the tool we'll be using to track
+changes in our data, code, and hyperparamters. With DVC, we can add some
+automation to the tuning process and be able to reproduce it all if we find a
+really good model.
+
+A few things DVC makes easier to do include:
+
+- Letting you make changes without worrying about finding them later
+- Onboarding other engineers to a project
+- Sharing experiments with other engineers on different machines
+- Tracking experiments and results in your Git repo and compare them from the
+  command line
+- Versioning the entire pipeline (including data, code, and commands run)
+
+For hyperparameter tuning, this means you can play with values and code changes
+without losing track of which changes made the best model and also have other
+engineers take a look. We'll do an example of this with grid search in DVC
+first.
+
+## Working with DVC
```

Comment:
I find this title confusing given the previous one: **Tuning with DVC**. Same thing given the context?

Reply:
I updated it a bit to hopefully clear that up!

Reply:
Like it 🙂

Reply:
Well... **Tuning with DVC** and then **Working with a DVC project** still sound pretty similar. It's not a big deal 💅 but if it was possible to compress that whole text in between and merge them, I'd consider it. Up to you